[Feature] 리프레시 토큰 기반 액세스 토큰 재발급 기능 추가

dbghwns123 opened the pull request:
## 📌 관련 이슈
- close #230 

## 📝 변경 사항
### AS-IS
- JWT 기반 인증은 AccessToken 단독 사용
- AccessToken 만료 시 재로그인 필요

### TO-BE
- 로그인 시 AccessToken + RefreshToken 함께 발급
- RefreshToken은 DB에 저장 및 재사용
- AccessToken 만료 시 RefreshToken으로 `/api/v1/auth/reissue` API 통해 AccessToken 재발급 가능

## 🔍 테스트
- [x] 테스트 코드 작성
- [x] API 테스트
- [x] 로컬 테스트

## 📸 스크린샷
- 처음 로그인 시 AccessToken, RefreshToken 모두 반환
![image](https://github.com/user-attachments/assets/5d48a828-2d21-4d3c-a608-9473f0f5de6a)

- 처음 발급된 AccessToken으로 예시 API 호출(사용자 권한이 있어야 하는 API) - 성공적으로 호출
![image](https://github.com/user-attachments/assets/68c7d07e-7194-4d26-8512-4639d6b0e374)

- 처음 발급된 AccessTokne의 토큰 만료 - 실패
![image](https://github.com/user-attachments/assets/a019ac39-6935-48c5-8b1c-d5a0252bb891)
![image](https://github.com/user-attachments/assets/ff7315d5-abd0-4721-98a9-88f672d156ec)

- 로그인시 발급된 RefreshToken으로 `/api/v1/auth/reissue` API 호출하여 새로운 AccessTokne 발급
![image](https://github.com/user-attachments/assets/eb852749-4bd8-4517-b52b-9e68a99031ee)

- 새로 발급받은 AccessToken으로 예시 API 호출(사용자 권한이 있어야 하는 API) - 성공적으로 호출
![image](https://github.com/user-attachments/assets/e6cca303-5675-4ba1-ac06-cffed6159e2a)


## ✅ 체크리스트
- [x] main/develop 브랜치가 아닌 feature 브랜치에서 작성했나요?
- [x] 코딩 컨벤션을 준수했나요?
- [x] 불필요한 코드는 제거했나요?
- [x] 주석은 충분히 작성했나요?
- [x] 테스트는 완료했나요?

## 🙋‍♂️ 리뷰어에게
- Admin 계정 포함 전체 유저 유형(Role)별 재발급 동작 확인 부탁드립니다.
- 이후 로그아웃 기능(RefreshToken 삭제)도 이어서 추가 예정입니다.

dbghwns123 committed:
feat:RefreshToken 엔티티 생성

dbghwns123 committed:
feat:RefreshTokenRepository 생성

dbghwns123 committed:
feat: AccessToken 및 RefreshToken 만료 시간 상수 추가

- AccessToken 유효기간: 15분(1000L * 60 * 15)
- RefreshToken 유효기간: 7일(1000L * 60 * 60 * 24 * 7)
- JWT 토큰 관리의 보안 및 사용성 개선 목적

dbghwns123 committed:
refactor: createToken 메서드명을 createAccessToken으로 변경 및 만료 시간 15분으로 단축

- 기존 1시간이던 accessToken 만료 시간을 15분으로 변경
- refreshToken 도입에 따라 accessToken 만료 시간 단축 적용
- 메서드 명칭을 createToken → createAccessToken으로 명확하게 변경

dbghwns123 committed:
feat:RefreshToken 발급 메서드 구현

dbghwns123 committed:
feat: LoginResponse에 refreshToken 필드 추가

dbghwns123 committed:
feat: 사용자 로그인 시 AccessToken과 RefreshToken 동시 발급 및 저장 기능 추가

dbghwns123 committed:
fix: RefreshToken 삭제 시 userId와 role을 함께 조건으로 사용하도록 수정

dbghwns123 committed:
feat: 호스트 로그인 시 AccessToken과 RefreshToken 동시 발급 및 저장 기능 추가

dbghwns123 committed:
feat: 관리자 로그인 시 AccessToken과 RefreshToken 동시 발급 및 저장 기능 추가

dbghwns123 committed:
feat: RefreshRequest record 생성 및 리프레시 토큰 유효성 검증 추가

dbghwns123 committed:
feat: RefreshResponse record 생성

dbghwns123 committed:
feat: 리프레시 토큰 기반 액세스 토큰 재발급 API 컨트롤러 추가

dbghwns123 committed:
feat:리프레시 토큰을 이용해 새로운 액세스 토큰을 재발급하는 서비스 로직 구현

dbghwns123 committed:
feat:리프레시 토큰 관련 ErrorCode 추가

dbghwns123 committed:
feat:SecurityConfig에 리프레시 토큰으로 새로운 액세스 토큰을 재발급하는 API를 인증없이 접근할 수 있도록 엔드포인트 추가

dbghwns123 committed:
test:정상적인 유저의 리프레시 토큰으로 새로운 액세스 토큰을 재발급에 성공 테스트 코드 작성

dbghwns123 committed:
test:존재하지 않는 리프레시 토큰으로 요청 시 예외가 발생하는 테스트 코드 작성

dbghwns123 committed:
test:DB 기준으로 만료된 리프레시 토큰이면 예외 발생 테스트 코드 작성

dbghwns123 committed:
test:리프레시 토큰 자체 유효성 검사 실패 - 예외 발생 테스트 코드 작성

JJae-story commented on src/main/java/com/meongnyangerang/meongnyangerang/service/AuthService.java:
```diff
@@ -90,4 +106,49 @@ public void checkNickname(String nickname) {
+    Long userId = tokenEntity.getUserId();
+    Role role = tokenEntity.getRole();
+
+    // 새로운 Access Token을 RefreshResponse로 감싸서 반환
+    return new RefreshResponse(switch (role) {
+      case ROLE_USER -> {
+        User user = userRepository.findById(userId)
+            .orElseThrow(() -> new MeongnyangerangException(NOT_EXIST_ACCOUNT));
+        if (user.getStatus() == UserStatus.DELETED) {
+          throw new MeongnyangerangException(ACCOUNT_DELETED);
+        }
+        yield jwtTokenProvider.createAccessToken(user.getId(), user.getEmail(), role.name(), user.getStatus());
+      }
+      case ROLE_HOST -> {
+        Host host = hostRepository.findById(userId)
+            .orElseThrow(() -> new MeongnyangerangException(NOT_EXIST_ACCOUNT));
+        if (host.getStatus() == HostStatus.DELETED || host.getStatus() == HostStatus.PENDING) {
+          throw new MeongnyangerangException(INVALID_AUTHORIZED);
+        }
+        yield jwtTokenProvider.createAccessToken(host.getId(), host.getEmail(), role.name(), host.getStatus());
+      }
+      case ROLE_ADMIN -> {
+        Admin admin = adminRepository.findById(userId)
+            .orElseThrow(() -> new MeongnyangerangException(NOT_EXIST_ACCOUNT));
+        yield jwtTokenProvider.createAccessToken(admin.getId(), admin.getEmail(), role.name(), admin.getStatus());
+      }
+    });
```
이 역할별 분기 로직을 각각의 메서드로 분리하면, 가독성과 유지보수가 용이해지지 않을까요? 어떻게 생각하시나요?

dbghwns123 replied:
의재님 피드백 반영해서 역할별 로직을 각각의 메서드로 분리했습니다. 전체 흐름이 확실히 더 읽기 좋아진 것 같아요. 감사합니다!

JJae-story reviewed (commented):
수고하셨습니다!

dbghwns123 committed:
refactor: RefreshToken 기반 AccessToken 재발급 로직 역할별 메서드 분리

- switch 식 내부 로직이 복잡해짐에 따라 역할별 분기 로직을 별도 메서드로 분리
- USER, HOST, ADMIN 각각의 예외 조건과 검증 로직을 메서드 단위로 분리하여 가독성과 유지보수성 향상
- 향후 역할 추가 시 확장성을 고려한 구조로 개선

Seokjin-K reviewed (approved):
수고하셨습니다!

JJae-story reviewed (approved):
수고하셨습니다!
